[metricbeat] meraki module with device_health metricset

tommyers-elastic opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Proposed commit message

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact

<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Beats.
-->

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
-

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/devices.go:
```diff
@@ -0,0 +1,231 @@
+func deviceDetailsToMapstr(details *meraki.ResponseItemOrganizationsGetOrganizationDevices) mapstr.M {
+	return mapstr.M{
+		"device.serial":       details.Serial,
+		"device.address":      details.Address,
```
Address and firmware are not added in fields.yml

tommyers-elastic replied:
good spot thanks

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/_meta/fields.yml:
```diff
@@ -0,0 +1,299 @@
+      fields:
+        - name: gateway
+          type: ip
+        - name: ip_status
```
I don't see any field ip_type getting the values. 
Is this meant to be ip_status?

tommyers-elastic replied:
great catch - thanks

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/_meta/fields.yml:
```diff
@@ -0,0 +1,299 @@
+        - name: head_license_id
+          type: keyword
+        - name: id
+          type: keyword
```
Keyword or long ?

tommyers-elastic replied:
thanks for checking - this is string data type in the code.

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/_meta/fields.yml:
```diff
@@ -0,0 +1,299 @@
+          type: long
+        - name: state
+          type: keyword
+        - name: total_duration_in_day
```
```suggestion
        - name: total_duration_in_days
```

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/switchports.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+func getDeviceSwitchports(client *meraki.Client, organizationID string, devices map[Serial]*Device, period time.Duration) error {
+	switches, res, err := client.Switch.GetOrganizationSwitchPortsBySwitch(organizationID, &meraki.GetOrganizationSwitchPortsBySwitchQueryParams{})
```
I am not aware about the Meraki service. So assuming all the API calls are correct. 
Looks like from the data.json of the metricset.

tommyers-elastic replied:
hope so 🙈

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/switchports.go:
```diff
@@ -0,0 +1,155 @@
+			metric := deviceDetailsToMapstr(device.details)
+
+			if switchport.port != nil {
+				metric["switch.port.id"] = switchport.port.PortID
```
Should we have these fields as switchport.* rather than switch.port.* ?

tommyers-elastic replied:
these are referred to a 'switch ports' in the [meraki docs](https://documentation.meraki.com/MS/Port_and_VLAN_Configuration/Switch_Ports), i think it's fine this way.

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/_meta/fields.yml:
```diff
@@ -0,0 +1,299 @@
+      type: group
+      fields:
+      - name: client_count
+        type: long
```
metric_type for this.

tommyers-elastic replied:
good spot - thanks

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/_meta/fields.yml:
```diff
@@ -0,0 +1,299 @@
+            type: keyword
+          - name: enabled
+            type: boolean
+          - name: config_overrides.allowed_vlans
```
Further group of config_overrides ?

tommyers-elastic replied:
yes good spot thanks

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/_meta/fields.yml:
```diff
@@ -0,0 +1,299 @@
+- name: device
```
Descriptions required for all the fields

tommyers-elastic replied:
that might come in the next release 🙃

tommyers-elastic replied:
or in the package

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/uplinks.go:
```diff
@@ -0,0 +1,172 @@
+							"uplink.latency.ms": dataPoint.LatencyMs,
+							"uplink.loss.pct":   dataPoint.LossPercent,
+
+							"device.serial":     uplink.lossAndLatency.Serial,    // _should_ be the same as `device.Serial`
```
Was this meant to be `device.serial` or `uplink.serial` ?

tommyers-elastic replied:
yeh this is device serial - the reason is that uplinks are not devices, but instead they are connections configured on devices. the unique identifier for an uplink is the device serial and the uplink interface e.g. `Q2BN-59GF-ZZEL/wan1`. this would be the primary WAN ( i.e. ethernet) on an MX device with the given serial number. 

some uplinks

ishleenk17 replied:
Do we add this field in field.yml ?

ishleenk17 commented on x-pack/metricbeat/module/meraki/device_health/device_health_test.go:
```diff
@@ -0,0 +1,63 @@
+package device_health
+
```
Add license hearders here.

ishleenk17 commented on x-pack/metricbeat/module/meraki/_meta/config.yml:
```diff
@@ -0,0 +1,6 @@
+- module: meraki
+  metricsets: ["device_health"]
+  enabled: true
+  period: 300s
```
Is a 5 minutes period sufficient for Meraki ?
What is the refresh interval internally for metrics of Meraki service ?

tommyers-elastic replied:
are you concerned about this being too long or too short?

i see uplink metrics in one minute buckets - the other metrics are hard to test because we don't have them in our test environment. but they appear default to 1 day (or longer) time periods in 10 minute buckets. there doesn't appear to be a limitation in how short the time period is for these metrics.

ishleenk17 replied:
I was concerned about it being long. 
For uplink, do you think, we should have shorter period ?

tommyers-elastic replied:
yeh, this is just a default though, don't need to worry about it now. you still get all the uplink metrics for the previous 5 minutes reported. so in the end you have one data point per minute. any longer than 5 minutes and you start to miss data points, which is why i included the upper bound on the period when the module starts up.

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

ishleenk17 commented:
Since, now we have decided to have fields mapping in Integrations only, we should be good till we are able to do data collect ion properly. 
In that respect changes look good!

tommyers-elastic commented on x-pack/metricbeat/module/meraki/device_health/_meta/fields.yml:
```diff
@@ -0,0 +1,307 @@
+    - name: license
+      type: group
+      fields:
+        - name: activation_date
```
good catch thanks

tommyers-elastic commented on x-pack/metricbeat/module/meraki/device_health/_meta/fields.yml:
```diff
@@ -0,0 +1,307 @@
+        type: keyword
+      - name: is_uplink
+        type: boolean
+      - name: power_usage_in_wh
```
i see [conventions](https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html#units) for common units in field names like `.ms` or `.pct` suffixes. i think a `.wh` might be confusing. also it's not a value value for a `units` tag in the mappings. i think i'd prefer to just keep it as is.

ishleenk17 commented:
The lint error is due to implicit memory aliasing within for loop.
Something like this might help resolving the error ?

    portCopy := port
    switchports = append(switchports, &switchport{port: &portCopy})

tommyers-elastic commented:
> The lint error is due to implicit memory aliasing within for loop. Something like this might help resolving the error ?
> 
> ```
> portCopy := port
> switchports = append(switchports, &switchport{port: &portCopy})
> ```

ah yeh thanks, i thought i fixed all these already. this issue actually has dissapeared in go1.22. i was gonna just put a global ignore rule for the linter error, but figured it might be easier to just guard against in the code. maybe i'll do both.

tommyers-elastic commented:
the more of these i fix the more the linter seems to find. since there's not really any need to do it, i reverted the fixes and added the exclude rule for the linkter.

cmacknz commented on go.mod:
```diff
@@ -1,6 +1,8 @@
-go 1.22
+go 1.22.0
+
+toolchain go1.22.7
```
```suggestion
toolchain go1.22.6
```

If we specify this it should match the one in .go-version (or it can just be removed) https://github.com/elastic/beats/blob/f4f1e3f9a4e81f41cba55939663cddc2cae67f02/.go-version#L1. Otherwise, CI is going to switch us to a Go 1.22.6 which will then download Go 1.22.7 on every build agent over and over again.

We should probably obsolete use of .go-version and gvm in favour of the toolchain directive but we haven't done that yet.

tommyers-elastic replied:
thanks for the detail here - i actually couldn't get this to build without having it here. i got errors about needing to commit the results of `go mod tidy`.

are you happy to approve once this is done? trying to avoid any more unnecessary long waits for CI :)

cmacknz reviewed (approved):
Approving go.mod changes.
